File: module2/class1/exemplo16.cpp
/*
================================================================================
ATIVIDADE PRÁTICA 16 - LOCK CONTENTION PERFORMANCE (C++)
================================================================================

OBJETIVO:
- Demonstrar degradação de performance devido à contention de locks
- Usar CPU profiler para identificar threads bloqueadas aguardando locks
- Otimizar usando fine-grained locking ou lock-free structures
- Medir impacto de lock contention na escalabilidade

PROBLEMA:
- Single global mutex causa serialização de threads
- Lock contention reduz paralelismo efetivo
- CPU Profiler mostrará threads blocked em mutex wait

SOLUÇÃO:
- Fine-grained locking ou partitioning
- Lock-free data structures quando possível

================================================================================
*/

#include <iostream>
#include <thread>
#include <mutex>
#include <vector>
#include <chrono>
#include <random>
using namespace std;

class CoarseGrainedCounter {
private:
    mutex globalMutex;
    vector<int> counters;
    
public:
    CoarseGrainedCounter(int size) : counters(size, 0) {}
    
    void increment(int index) {
        // PERFORMANCE ISSUE: Single global lock for all operations
        lock_guard<mutex> lock(globalMutex); // All threads contend for same lock
        
        counters[index]++;
        
        // Simulate some work while holding the lock
        this_thread::sleep_for(chrono::microseconds(1));
    }
    
    int getSum() {
        lock_guard<mutex> lock(globalMutex);
        int sum = 0;
        for (int val : counters) {
            sum += val;
        }
        return sum;
    }
};

void demonstrateLockContention() {
    cout << "Starting lock contention demonstration..." << endl;
    cout << "Monitor CPU profiler - should see threads blocked on mutex wait" << endl;
    
    const int NUM_THREADS = 8;
    const int OPERATIONS_PER_THREAD = 1000;
    const int COUNTER_SIZE = 100;
    
    CoarseGrainedCounter counter(COUNTER_SIZE);
    vector<thread> threads;
    
    random_device rd;
    mt19937 gen(rd());
    uniform_int_distribution<> indexDist(0, COUNTER_SIZE - 1);
    
    auto start = chrono::high_resolution_clock::now();
    
    // Launch threads that will contend for the same global lock
    for (int t = 0; t < NUM_THREADS; t++) {
        threads.emplace_back([&counter, &indexDist, OPERATIONS_PER_THREAD, t]() {
            mt19937 localGen(t); // Thread-local generator
            uniform_int_distribution<> localDist(0, 99);
            
            for (int i = 0; i < OPERATIONS_PER_THREAD; i++) {
                int index = localDist(localGen);
                counter.increment(index); // All threads compete for global lock
                
                if (i % 200 == 0) {
                    cout << "Thread " << t << " completed " << i << "/" << OPERATIONS_PER_THREAD << " operations" << endl;
                }
            }
        });
    }
    
    // Wait for all threads
    for (auto& t : threads) {
        t.join();
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << "Lock contention test completed in: " << duration.count() << " ms" << endl;
    cout << "Total sum: " << counter.getSum() << endl;
    cout << "Threads: " << NUM_THREADS << ", Operations per thread: " << OPERATIONS_PER_THREAD << endl;
    cout << "Lock contention severely limited parallelism" << endl;
}

int main() {
    cout << "Starting lock contention demonstration..." << endl;
    cout << "Task: Multiple threads incrementing counters with global lock" << endl;
    cout << "Monitor CPU Usage Tool for lock contention and thread blocking" << endl << endl;
    
    demonstrateLockContention();
    
    cout << endl << "=== PROFILING ANALYSIS ===" << endl;
    cout << "Check CPU profiler for:" << endl;
    cout << "- Threads blocked waiting for mutex" << endl;
    cout << "- Low CPU utilization due to serialization" << endl;
    cout << "- Lock contention overhead" << endl;
    
    return 0;
}

/*
================================================================================
VERSÃO CORRIGIDA (DESCOMENTE PARA USAR FINE-GRAINED LOCKING)
================================================================================

#include <iostream>
#include <thread>
#include <mutex>
#include <vector>
#include <chrono>
#include <random>
using namespace std;

class FineGrainedCounter {
private:
    vector<mutex> mutexes;  // One mutex per counter
    vector<int> counters;
    
public:
    FineGrainedCounter(int size) : mutexes(size), counters(size, 0) {}
    
    void increment(int index) {
        // CORREÇÃO: Lock only the specific counter, not everything
        lock_guard<mutex> lock(mutexes[index]); // Fine-grained locking
        
        counters[index]++;
        
        // Same work, but now only blocks access to this specific counter
        this_thread::sleep_for(chrono::microseconds(1));
    }
    
    int getSum() {
        // CORREÇÃO: Lock all mutexes in deterministic order to avoid deadlock
        vector<unique_lock<mutex>> locks;
        for (auto& m : mutexes) {
            locks.emplace_back(m);
        }
        
        int sum = 0;
        for (int val : counters) {
            sum += val;
        }
        return sum;
    }
};

// Alternative: Lock-free approach using atomic operations
class LockFreeCounter {
private:
    vector<atomic<int>> counters;
    
public:
    LockFreeCounter(int size) : counters(size) {
        for (auto& counter : counters) {
            counter = 0;
        }
    }
    
    void increment(int index) {
        // CORREÇÃO: Atomic increment - no locks needed
        counters[index].fetch_add(1, memory_order_relaxed);
        
        // Simulate work without holding any locks
        this_thread::sleep_for(chrono::microseconds(1));
    }
    
    int getSum() {
        int sum = 0;
        for (const auto& counter : counters) {
            sum += counter.load(memory_order_relaxed);
        }
        return sum;
    }
};

void demonstrateFineGrainedLocking() {
    cout << "Starting fine-grained locking demonstration..." << endl;
    cout << "Monitor CPU profiler - should see improved parallelism" << endl;
    
    const int NUM_THREADS = 8;
    const int OPERATIONS_PER_THREAD = 1000;
    const int COUNTER_SIZE = 100;
    
    FineGrainedCounter counter(COUNTER_SIZE);
    vector<thread> threads;
    
    auto start = chrono::high_resolution_clock::now();
    
    // Launch threads - now they can work in parallel on different counters
    for (int t = 0; t < NUM_THREADS; t++) {
        threads.emplace_back([&counter, OPERATIONS_PER_THREAD, COUNTER_SIZE, t]() {
            mt19937 localGen(t);
            uniform_int_distribution<> localDist(0, COUNTER_SIZE - 1);
            
            for (int i = 0; i < OPERATIONS_PER_THREAD; i++) {
                int index = localDist(localGen);
                counter.increment(index); // Much less contention
            }
        });
    }
    
    for (auto& t : threads) {
        t.join();
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << "Fine-grained locking completed in: " << duration.count() << " ms" << endl;
    cout << "Total sum: " << counter.getSum() << endl;
}

void demonstrateLockFree() {
    cout << "Starting lock-free demonstration..." << endl;
    
    const int NUM_THREADS = 8;
    const int OPERATIONS_PER_THREAD = 1000;
    const int COUNTER_SIZE = 100;
    
    LockFreeCounter counter(COUNTER_SIZE);
    vector<thread> threads;
    
    auto start = chrono::high_resolution_clock::now();
    
    // Launch threads - completely lock-free operation
    for (int t = 0; t < NUM_THREADS; t++) {
        threads.emplace_back([&counter, OPERATIONS_PER_THREAD, COUNTER_SIZE, t]() {
            mt19937 localGen(t);
            uniform_int_distribution<> localDist(0, COUNTER_SIZE - 1);
            
            for (int i = 0; i < OPERATIONS_PER_THREAD; i++) {
                int index = localDist(localGen);
                counter.increment(index); // No locks at all!
            }
        });
    }
    
    for (auto& t : threads) {
        t.join();
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << "Lock-free approach completed in: " << duration.count() << " ms" << endl;
    cout << "Total sum: " << counter.getSum() << endl;
}

int main() {
    cout << "Starting optimized locking demonstration..." << endl;
    cout << "Task: Comparing different locking strategies" << endl;
    cout << "Monitor CPU Usage Tool for improved parallelism" << endl << endl;
    
    demonstrateFineGrainedLocking();
    cout << endl;
    demonstrateLockFree();
    
    cout << endl << "=== OPTIMIZATION RESULTS ===" << endl;
    cout << "Improvements:" << endl;
    cout << "- Fine-grained locking reduces contention" << endl;
    cout << "- Lock-free atomics eliminate blocking entirely" << endl;
    cout << "- Much better thread utilization and scalability" << endl;
    cout << "- Significantly faster execution" << endl;
    
    return 0;
}

================================================================================
*/

File: module2/class1/exemplo16.cs
/*
================================================================================
ATIVIDADE PRÁTICA 16 - THREAD POOL STARVATION (C#)
================================================================================

OBJETIVO:
- Demonstrar thread pool starvation com blocking operations
- Usar CPU/Thread profiler para identificar thread pool exhaustion
- Otimizar usando async/await para non-blocking operations
- Medir impacto de thread pool starvation na escalabilidade

PROBLEMA:
- Blocking I/O operations consume thread pool threads
- Thread pool starvation impede novas tasks de executar
- Thread profiler mostrará all threads blocked

SOLUÇÃO:
- Usar async I/O para liberar threads durante waits
- ConfigureAwait(false) para library code

================================================================================
*/

using System;
using System.Threading;
using System.Threading.Tasks;
using System.Diagnostics;
using System.Net.Http;

class Program {
    static void DemonstrateThreadPoolStarvation() {
        Console.WriteLine("Starting thread pool starvation demonstration...");
        Console.WriteLine("Monitor thread profiler - should see thread pool exhaustion");
        
        const int CONCURRENT_OPERATIONS = 50;
        int initialThreads = ThreadPool.ThreadCount;
        
        var sw = Stopwatch.StartNew();
        
        // PERFORMANCE ISSUE: Blocking operations that tie up thread pool threads
        var tasks = new Task[CONCURRENT_OPERATIONS];
        
        for (int i = 0; i < CONCURRENT_OPERATIONS; i++) {
            int taskId = i;
            tasks[i] = Task.Run(() => {
                // PERFORMANCE ISSUE: Synchronous blocking call ties up thread
                Thread.Sleep(2000); // Blocking sleep consumes thread pool thread
                
                Console.WriteLine($"Task {taskId} completed on thread {Thread.CurrentThread.ManagedThreadId}");
                return taskId;
            });
        }
        
        // Wait for all tasks - but thread pool may be starved
        Task.WaitAll(tasks);
        
        sw.Stop();
        
        Console.WriteLine($"Thread pool starvation test completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Initial threads: {initialThreads}");
        Console.WriteLine($"Final thread count: {ThreadPool.ThreadCount}");
        Console.WriteLine($"Tasks executed: {CONCURRENT_OPERATIONS}");
        Console.WriteLine("All thread pool threads were blocked during wait periods");
    }
    
    static void Main() {
        // Configure thread pool to show starvation more clearly
        ThreadPool.SetMaxThreads(10, 10);
        ThreadPool.SetMinThreads(2, 2);
        
        Console.WriteLine("Starting thread pool starvation demonstration...");
        Console.WriteLine("Task: Running many blocking operations concurrently");
        Console.WriteLine("Monitor Thread Usage and CPU profiler for thread pool issues");
        Console.WriteLine();
        
        DemonstrateThreadPoolStarvation();
        
        Console.WriteLine();
        Console.WriteLine("=== PROFILING ANALYSIS ===");
        Console.WriteLine("Check profiler for:");
        Console.WriteLine("- Thread pool thread exhaustion");
        Console.WriteLine("- Blocked threads during I/O waits");
        Console.WriteLine("- Poor thread utilization");
    }
}

/*
================================================================================
VERSÃO CORRIGIDA (DESCOMENTE PARA USAR ASYNC/AWAIT)
================================================================================

using System;
using System.Threading;
using System.Threading.Tasks;
using System.Diagnostics;

class Program {
    static async Task DemonstrateAsyncOperations() {
        Console.WriteLine("Starting async operations demonstration...");
        Console.WriteLine("Monitor thread profiler - should see efficient thread usage");
        
        const int CONCURRENT_OPERATIONS = 50;
        int initialThreads = ThreadPool.ThreadCount;
        
        var sw = Stopwatch.StartNew();
        
        // CORREÇÃO: Async operations that release threads during waits
        var tasks = new Task<int>[CONCURRENT_OPERATIONS];
        
        for (int i = 0; i < CONCURRENT_OPERATIONS; i++) {
            int taskId = i;
            tasks[i] = PerformAsyncOperation(taskId);
        }
        
        // CORREÇÃO: Async wait that doesn't block threads
        var results = await Task.WhenAll(tasks);
        
        sw.Stop();
        
        Console.WriteLine($"Async operations completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Initial threads: {initialThreads}");
        Console.WriteLine($"Final thread count: {ThreadPool.ThreadCount}");
        Console.WriteLine($"Tasks executed: {CONCURRENT_OPERATIONS}");
        Console.WriteLine("Threads were returned to pool during async waits");
    }
    
    static async Task<int> PerformAsyncOperation(int taskId) {
        // CORREÇÃO: Async delay releases the thread back to pool
        await Task.Delay(2000).ConfigureAwait(false); // Non-blocking wait
        
        Console.WriteLine($"Task {taskId} completed on thread {Thread.CurrentThread.ManagedThreadId}");
        return taskId;
    }
    
    // Example with real async I/O
    static async Task DemonstrateAsyncIO() {
        Console.WriteLine("Starting async I/O demonstration...");
        
        const int CONCURRENT_REQUESTS = 20;
        using var httpClient = new HttpClient();
        
        var sw = Stopwatch.StartNew();
        
        // CORREÇÃO: Async HTTP calls that don't block threads
        var tasks = new Task<string>[CONCURRENT_REQUESTS];
        
        for (int i = 0; i < CONCURRENT_REQUESTS; i++) {
            tasks[i] = httpClient.GetStringAsync($"https://httpbin.org/delay/1");
        }
        
        var responses = await Task.WhenAll(tasks);
        
        sw.Stop();
        
        Console.WriteLine($"Async I/O completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Concurrent requests: {CONCURRENT_REQUESTS}");
        Console.WriteLine($"Average response length: {responses.Average(r => r.Length):F0} chars");
        Console.WriteLine("HTTP I/O was non-blocking - threads returned to pool");
    }
    
    static async Task Main() {
        // Configure thread pool to show the difference
        ThreadPool.SetMaxThreads(10, 10);
        ThreadPool.SetMinThreads(2, 2);
        
        Console.WriteLine("Starting optimized async demonstration...");
        Console.WriteLine("Task: Using async/await for concurrent operations");
        Console.WriteLine("Monitor Thread Usage for efficient thread utilization");
        Console.WriteLine();
        
        await DemonstrateAsyncOperations();
        Console.WriteLine();
        await DemonstrateAsyncIO();
        
        Console.WriteLine();
        Console.WriteLine("=== OPTIMIZATION RESULTS ===");
        Console.WriteLine("Improvements:");
        Console.WriteLine("- Threads returned to pool during async waits");
        Console.WriteLine("- No thread pool starvation");
        Console.WriteLine("- Better scalability for I/O-bound operations");
        Console.WriteLine("- More efficient use of system resources");
    }
}

================================================================================
*/
